demos-europe/demosplan-core
# feat: implement flysystem for S3

### Ticket
DPLAN-3415
or
https://www.dev.diplanung.de/DefaultCollection/EfA%20DiPlanung/_workitems/edit/8179

This PR documents implements Flysystem in dplan as a flexible adapter for file actions. Initially there are two storages: local and s3. You may switch  them by setting `FILES_SOURCE=s3.storage` or `FILES_SOURCE=local.storage` respectively in your .env.local.
To enable local s3 buckets you need to enable minio with `containers.minio.enable: true` and `containers.minio_mc.enable: true` in your `.defaults.yml`.

To check the local s3 files you may use the minio ui at `http://localhost:9005`. You can find the login in your docker-compose.yml. You need to create a bucket called `dplan-files` in the ui

I tried to keep this PR as small as possible, but as every file related action needs to be checked and eventually adjusted it is quite large. I am happy to assist when any questions raise.


### How to review/test
You may first run your local instance with the default `FILES_SOURCE=local.storage` and check that any existing files can still be found and any file related action works. You may then copy your local files to the local s3 bucket with `bin/<project> dplan:files:move local s3 --no-delete` and change `FILES_SOURCE=s3.storage` and test again that any file is found and any file related action works





### PR Checklist


Delete the checkbox if it doesn't apply/isn't necessary.
